Add methods for LLVMTargetDataRef

yowl opened the pull request:
This adds methods to `LLVMTargetRef` for some of the underlying LLVM api calls, covers the ones I want, struct element offsets, and others that I saw that took a `LLVMTargetRef` as the first parameter.  I also added a class that wraps the `LLVMTargetRef` struct, `LLVMTargetData` thinking of #140 but not sure I've understood that right as doesn't seem to add much so let me know if that should come out.  Added some simple tests for the new methods in `LLVMTargetRef` to at least make sure they return something sensible.

tannergooding commented on Tests/LLVMSharp.UnitTests/TargetData.cs:
```diff
@@ -0,0 +1,69 @@
+using LLVMSharp.Interop;
```
nit: This needs a license header like the other files have:
> // Copyright (c) Microsoft and Contributors. All rights reserved. Licensed under the University of Illinois/NCSA Open Source License. See LICENSE.txt in the project root for license information.

tannergooding replied:
nit: This file was adder to a folder name `Tests` rather than `tests`.

Git is case sensitive and so the casing needs to be adjusted to match the existing `tests`. You'll likely need to `git mv src dst`, possibly to some other folder/name and then to the correctly cased folder/name

tannergooding commented on sources/LLVMSharp/LLVMTargetData.cs:
```diff
@@ -0,0 +1,59 @@
+using LLVMSharp.Interop;
```
nit: This needs a license header like the other files have:
> // Copyright (c) Microsoft and Contributors. All rights reserved. Licensed under the University of Illinois/NCSA Open Source License. See LICENSE.txt in the project root for license information.

yowl replied:
Thanks, I made the same mistake last time I think....

tannergooding commented on sources/LLVMSharp/LLVMTargetData.cs:
```diff
@@ -0,0 +1,59 @@
+
+namespace LLVMSharp
+{
+    public class LLVMTargetData
```
Like I mentioned in #140, `LLVMTargetDataRef` is "basically" a reference to a `llvm:DataLayout`. So for the "managed" API, this should likely be called `DataLayout`.

The class should likewise be `sealed` (as going from sealed to unsealed is non-breaking; but going from unsealed to sealed is), expose the underlying `LLVMTargetDataRef` as a `Handle` property and implement `IEquatable<DataLayout>` (similarly to the base APIs exposed on https://github.com/microsoft/LLVMSharp/blob/master/sources/LLVMSharp/IRBuilderBase.cs).

The public constructor, rather than taking a `LLVMTargetDataRef` directly should likely take a `string` and internally call `LLVMCreateTargetData`.
Eventually there would also be an internal constructor that is created from `Module.DataLayout` (which doesn't exist today, but which would be similar to the [IntegerType](https://github.com/microsoft/LLVMSharp/blob/master/sources/LLVMSharp/Types/IntegerType.cs) constructor)

I believe `LLVMCreateTargetData` always creates a new `DataLayout` and so there wouldn't be any need to cache instances like happens with https://github.com/microsoft/LLVMSharp/blob/master/sources/LLVMSharp/LLVMContext.cs#L12, but that would be needed once `Module.DataLayout` exists.

Finally, the exposed method names would ideally match, as nearly as possible, the name and shape of the APIs exposed by `llvm::DataLayout`. https://llvm.org/doxygen/group__LLVMCTarget.html#ga0b1cfce31433d9948825ed2e2e203830 lists the method that is used behind the scenes.

For example, [OffsetOfElement](https://llvm.org/doxygen/group__LLVMCTarget.html#ga0b1cfce31433d9948825ed2e2e203830) isn't actually part of `DataLayout`, it is part of `llvm:StructLayout` which is a `DataLayout` like type which `DataLayout` can create (it is created via [getStructLayout](https://llvm.org/doxygen/classllvm_1_1DataLayout.html#a61135fb8666f0b7a37b4e1bbcf1db131) which itself takes a `StructType`: https://github.com/microsoft/LLVMSharp/blob/master/sources/LLVMSharp/Types/StructType.cs).
* There are basically two options here, given that there isn't a `StructLayout` handle exposed via the C API
  * The first would be to expose this directly on `DataLayout` as `getElementOffset(StructType type, uint idx)`
  * The second would be to create a `StructLayout` struct which wraps a `StructType` and a `DataLayout`, this more closely "mirrors" the C++ API

[SizeOfTypeInBits](https://llvm.org/doxygen/group__LLVMCTarget.html#gaa29cd97b498d32b9c5663574bd2659d7) on the other hand is part of `llvm::DataLayout` and corresponds to a method name [getTypeSizeInBits](https://llvm.org/doxygen/classllvm_1_1DataLayout.html#acf0b1898efd7f81a078e9288befd9290).

That API would take a `LLVMSharp.Type` rather than a TypeRef, that way it is also using the managed API.

yowl replied:
Thanks for the help and detailed reply.  Let me digest and see if I can get it right.

yowl commented:
I've made some changes in line with the feedback, hopefully I'm getting closer.  I went for option 2 and created a `StructLayout` class.  It takes both a `DataLayout` and the `StructType` as it needs access to the `LLVMTargetDataRef` to call down.  I dropped one of the methods `CallFrameAlignmentOfType`, as it ends up calling `getABITypeAlignment` as does `ABIAlignmentOfType`

```
 unsigned LLVMABIAlignmentOfType(LLVMTargetDataRef TD, LLVMTypeRef Ty) {
   return unwrap(TD)->getABITypeAlignment(unwrap(Ty));
 }
 
 unsigned LLVMCallFrameAlignmentOfType(LLVMTargetDataRef TD, LLVMTypeRef Ty) {
   return unwrap(TD)->getABITypeAlignment(unwrap(Ty));
 }
```
So in order to keep the name in line with the c name, I just dropped the duplicate.

tannergooding commented on sources/LLVMSharp/DataLayout.cs:
```diff
@@ -0,0 +1,65 @@
+
+        public LLVMTargetDataRef Handle { get; }
+
+        public StructLayout CreateStructLayout(StructType structType)
```
```suggestion
        public StructLayout GetStructLayout(StructType structType)
```

To match the C++ name: https://llvm.org/doxygen/classllvm_1_1DataLayout.html#a61135fb8666f0b7a37b4e1bbcf1db131

tannergooding commented on sources/LLVMSharp/DataLayout.cs:
```diff
@@ -0,0 +1,65 @@
+            return new StructLayout(this, structType);
+        }
+
+        public ulong TypeSizeInBits(Type type)
```
Likewise, prefix with `Get` and similar for others below where it is a non-property and the C++ name is `get*`
```suggestion
        public ulong GetTypeSizeInBits(Type type)
```

tannergooding commented:
Overall this LGTM now, minus the naming nits on methods.

yowl commented:
Thanks!

tannergooding commented:
Just wanted to make sure you saw (https://github.com/microsoft/LLVMSharp/pull/141#discussion_r478564489):
> Likewise, prefix with Get and similar for others below where it is a non-property and the C++ name is get*

I didn't explicitly provide a suggestion for all the names that should be updated and it looks like a couple still remain. https://llvm.org/doxygen/classllvm_1_1DataLayout.html has the C++ names if you need a reference.

yowl commented:
Thanks, I'll look at the others.

tannergooding commented:
Thanks!
